File: custisinstall/replicate.pl
#!/usr/bin/perl
# OBSOLETE: better use PHP version "replicate.php".

# A script for MediaWiki4Intranet page replication, automatically
# replicating used images and templates, and supporting incremental
# replication.
# Скрипт для репликации вики-страниц между разными MediaWiki,
# поддерживает автоматическую репликацию использованных изображений
# и шаблонов, и инкрементальную репликацию.

# REQUIRES modified MediaWiki import/export mechanism, see MediaWiki4Intranet patch:
# http://wiki.4intra.net/MW_Import_Export

# ТРЕБУЕТ модифицированного механизма импорта/экспорта MediaWiki, см. патч MediaWiki4Intranet:
# http://wiki.4intra.net/MW_Import_Export

use strict;
use constant HELP_TEXT => <<EOF;
MediaWiki4Intranet replication script
Copyright (c) 2010+ Vitaliy Filippov <vitalif\@mail.ru>

USAGE: $0 [OPTIONS] <replication-config.ini> [targets...]

OPTIONS:

-t HOURS
  only select pages which were changed during last HOURS hours.
  I.e. if the replication script is ran each day, you can specify -t 24 to
  export only pages changed since last run, or better -t 25 to allow some
  overlap with previous day and make replication more reliable.

-t 'YYYY-MM-DD[ HH:MM:SS]'
  same as above, but specify date/time, not the relative period in hours.

-i
  when using regular incremental replication (-t option), the following
  situation may be possible:
  * template was created, say, on 2011-10-11
  * it is outside replication category, therefore does not replicate by itself
    (-t 24 is used each day)
  * article was created, say, on 2011-10-13, in replication category
  * so article replicates 2011-10-14, but the template does not
    (because it's not modified during last 24 hours)
  This replication script by default ignores last modification date for
  templates and images. You can change this behaviour using this -i option.

When called without target list, $0 will attempt to replicate all targets
found in config file. There must be 2 sections in config file according to
each target and named "<Target>SourceWiki" and "<Target>DestinationWiki".

Config file fragment syntax (Replace __Test__ with desired [target] name):

[__Test__SourceWiki]
URL=<source wiki url>
Category=<source category name for selecting pages>
NotCategory=<source category name for replication denial>
RemoveConfidential=<'yes' or 'no' (default)>
FullHistory=<'yes' or 'no' (default), 'yes' replicates all page revisions, not only the last one>
BasicLogin=<HTTP basic auth username, if needed>
BasicPassword=<HTTP basic auth password, if needed>

[__Test__DestinationWiki]
URL=<destination wiki url>
BasicLogin=<HTTP basic auth username, if needed>
BasicPassword=<HTTP basic auth password, if needed>
User=<name of a user having import rights in destination wiki>
Password=<his password>
EOF

BEGIN
{
    require File::Basename;
    my ($a) = $0 =~ /^(.*)$/iso;
    chdir(File::Basename::dirname($a));
}

use lib qw(.);
use URI;
use POSIX qw(strftime);
use File::Path 2.07;
use File::Temp;
use Time::HiRes qw(CLOCK_REALTIME clock_gettime);
use LWP::UserAgent;
use Digest::SHA1;

use MIME::Base64;
use URI::Escape;
use HTTP::Cookies;
use HTTP::Date;
use HTTP::Request::Common;
use HTML::Entities;

my $BUFSIZE = 0x10000;

my $since_time;
my $ignore_since_images = 1;
my $config_file;
my @targets;

for (my $i = 0; $i < @ARGV; $i++)
{
    if ($ARGV[$i] eq '-t')
    {
        $since_time = $ARGV[++$i];
        if ($since_time !~ /^\s*(\d{4,}-\d{2}-\d{2}(?:\s+\d{2}:\d{2}:\d{2})?)\s*$/s)
        {
            $since_time = int(time - 3600*$since_time);
            $since_time = strftime("%Y-%m-%d %H:%M:%S", localtime($since_time));
        }
        else
        {
            $since_time = $1;
        }
    }
    elsif ($ARGV[$i] eq '-i')
    {
        $ignore_since_images = 0;
    }
    elsif (!defined $config_file)
    {
        $config_file = $ARGV[$i];
    }
    else
    {
        push @targets, lc $ARGV[$i];
    }
}

my $config = read_config($config_file) || die HELP_TEXT();

$| = 1;

my $replicating;
sub logp { strftime("[%Y-%m-%d %H:%M:%S] [$replicating]", localtime) }

my $log;
@targets = keys %$config unless @targets;
for (@targets)
{
    # Replication targets
    $replicating = $_;
    print logp()." Begin replication\n";
    eval { replicate($config->{$_}->{src}, $config->{$_}->{dest}) };
    print STDERR logp()." Could not replicate:\n$@" if $@;
}

# Login into wiki described by $params with LWP user agent $ua,
# $msgdesc is the wiki description for log messages
sub login_into
{
    my ($ua, $params, $msgdesc) = @_;
    my $uri = URI->new($params->{url})->canonical;
    $ua->credentials($uri->host_port, undef, $params->{basiclogin} || undef, $params->{basicpassword});
    if ($params->{user} && $params->{password})
    {
        my $response = $ua->request(GET("$params->{url}/index.php?title=Special:UserLogin"));
        die logp()." Could not retrieve login form from the $msgdesc: ".$response->status_line
            unless $response->code == 200;
        my ($token) = $response->content =~ /<input[^<>]*name="wpLoginToken"[^<>]*value="([^"]+)"[^<>]*>/so;
        $response = $ua->request(POST("$params->{url}/index.php?title=Special:UserLogin&action=submitlogin&type=login",
            Content => [
                wpName         => $params->{user},
                wpPassword     => $params->{password},
                wpLoginAttempt => 1,
                wpLoginToken   => $token,
            ],
        ));
        die logp()." Could not login into $msgdesc under user '$params->{user}': ".$response->status_line
            unless $response->code == 302;
    }
}

# Generate export page list from wiki $url using $params and $desc as error description
sub page_list_load
{
    my ($ua, $url, $desc, $params) = @_;
    my $response = $ua->request(
        POST "$url/index.php?title=Special:Export&action=submit",
        [
            addcat => "Добавить",
            @$params
        ],
    );
    unless ($response->is_success)
    {
        die logp()." Could not retrieve page list from $desc ".
            "($url/index.php?title=Special:Export&action=submit): ".
            $response->status_line;
    }
    my $text = $response->content;
    ($text) = $text =~ m!<textarea[^<>]*>(.*?)</textarea>!iso;
    $text =~ s/^\s*//so;
    $text =~ s/\s*$//so;
    decode_entities($text);
    return $text;
}

# Retrieve list of Wiki pages from category $cat,
# NOT in category $notcat, with all used images and
# templates by default, but only modified after $modifydate
sub page_list
{
    my ($ua, $src, $cat, $notcat, $modifydate, $ignore_since_images) = @_;
    $cat ||= '';
    $notcat ||= '';
    $modifydate ||= '';
    $ignore_since_images = $ignore_since_images && $modifydate ne '';
    my $desc = "Category:$cat";
    $desc .= " MINUS category:$notcat" if $notcat ne '';
    $desc .= ", modified after $modifydate" if $modifydate ne '';
    $desc .= ", with all used images/templates" if !$ignore_since_images;
    my $text = page_list_load($ua, $src->{url}, $desc, [
        catname     => $cat,
        notcategory => $notcat,
        modifydate  => $modifydate,
        ($ignore_since_images ? () : (
            templates => 1,
            images    => 1,
        ))
    ]);
    if (!$text)
    {
        print logp()." No pages need replication in $desc\n";
    }
    elsif ($ignore_since_images)
    {
        # Add templates and images in a separate request, without passing modifydate
        $text = page_list_load($ua, $src->{url}, $desc, [
            notcategory => ($notcat ||= ''),
            templates   => 1,
            images      => 1,
            pages       => $text,
        ]);
    }
    return $text;
}

sub replicate
{
    my ($src, $dest) = @_;
    my $ua = MYLWPUserAgent->new;
    $ua->cookie_jar(HTTP::Cookies->new);
    my ($uri, $response);
    # Login into source wiki
    login_into($ua, $src, 'source wiki');
    # Read page list for replication
    my $text = page_list($ua, $src, $src->{category}, $src->{notcategory}, $since_time, $ignore_since_images);
    if (!$text)
    {
        # No pages
        return 1;
    }
    my $ts = clock_gettime(CLOCK_REALTIME);
    # Read export XML / multipart file
    my $fh = File::Temp->new;
    my $fn = $fh->filename;
    my $auth;
    $auth = 'Basic '.encode_base64($src->{basiclogin}.':'.$src->{basicpassword}) if $src->{basiclogin};
    $response = $ua->request(
        POST("$src->{url}/index.php?title=Special:Export&action=submit", [
            images        => 1,
            selfcontained => 1,
            wpDownload    => 1,
            pages         => $text,
            curonly       => 1,
            ($src->{removeconfidential} ? () : (confidential => 1)),
            ($src->{fullhistory}        ? () : (curonly => 1)),
        ]),
        $fn, # Let LWP::UserAgent write response content into this file
    );
    die logp()." Could not retrieve export XML file from '$src->{url}/index.php?title=Special:Export&action=submit': ".$response->status_line
        unless $response->is_success;
    my $tx = clock_gettime(CLOCK_REALTIME);
    print logp().sprintf(" Retrieved %d bytes in %.2f seconds\n", -s $fn, $tx-$ts);
    # Login into destination wiki
    login_into($ua, $dest, 'destination wiki');
    # Retrieve token for importing
    $response = $ua->request(GET "$dest->{url}/index.php?title=Special:Import");
    die logp()." Could not retrieve Special:Import page from '$dest->{url}/index.php?title=Special:Import: ".$response->status_line
        unless $response->is_success;
    $text = $response->content;
    my $token = $text =~ /<input([^<>]*name="editToken"[^<>]*)>/iso &&
        $1 =~ /value=\"([^\"]*)\"/iso && $1 || undef;
    # Run import
    $response = $ua->request(POST("$dest->{url}/index.php?title=Special:Import&action=submit",
        Content_Type => 'form-data',
        Content      => [
            source    => 'upload',
            editToken => $token,
            xmlimport => [ $fn ],
        ],
    ));
    die logp()." Could not import into '$dest->{url}/index.php?title=Special:Import&action=submit': ".$response->status_line
        unless $response->is_success;
    die logp()." Could not import into $dest->{url}: $1" if $response->content =~ /<p[^<>]*class\s*=\s*["']?error[^<>]*>\s*(.*?)\s*<\/p\s*>/iso;
    my $tp = clock_gettime(CLOCK_REALTIME);
    print sprintf(logp()." Imported in %.2f seconds\n", $tp-$tx);
    $text = $response->content;
    # Extract the import report
    my ($report) = $text =~ /<!--\s*start\s*content\s*-->.*?<ul>(.*?)<\/ul>/iso;
    for ($report)
    {
        s/&nbsp;/ /giso;
        s/\s+/ /giso;
        s/<li[^<>]*>/\n/giso;
        s/<\/?[a-z0-9_:\-]+(\/?\s+[^<>]*)?>//giso;
        decode_entities($_);
        s/^\s+//so;
        s/\s+$//so;
    }
    die logp()." Could not replicate, no import report found in response content:\n$text\n"
        unless $report;
    print logp()." Report:\n$report\n";
    # Finished
    1;
}

sub read_config
{
    my ($file) = @_;
    my $fh;
    return undef unless open $fh, "<$file";
    my $cfg = {};
    my $target;
    my $th = {};
    my ($key, $h, $k, $v);
    while (<$fh>)
    {
        s/\s+$//so;
        s/(^|\s+)(;|\#).*$//so;
        next unless $_;
        if (/^\s*\[([^\]]*)(Source|Destination)Wiki\]\s*$/iso)
        {
            $target = lc $1;
            $key = lc $2 eq 'source' ? 'src' : 'dest';
            $th->{$target} = 1 if $cfg->{$target}->{lc $2 eq 'source' ? 'dest' : 'src'};
        }
        elsif (/^\s*([^=]*[^\s=])\s*=\s*(.*)/iso)
        {
            $k = lc $1;
            $v = $2;
            $h = $cfg;
            $h = ($h->{$target}->{$key} ||= {}) if $target && $key;
            if ($k eq 'url')
            {
                $v =~ s!/+$!!so;
            }
            elsif ($k eq 'fullhistory' || $k eq 'removeconfidential')
            {
                $v = lc $v;
                $v = $v eq 'yes' || $v eq 'true' || $v eq 'on' || $v eq '1' ? 1 : 0;
            }
            $h->{$k} = $v;
        }
    }
    close $fh;
    for (keys %$cfg)
    {
        delete $cfg->{$_} unless $th->{$_};
    }
    return undef unless keys %$cfg;
    return $cfg;
}

sub sha1_file
{
    my ($fn) = @_;
    my $hash = '';
    if (open FH, "<$fn")
    {
        $hash = Digest::SHA1->new;
        $hash->addfile(*FH);
        $hash = $hash->hexdigest;
        close FH;
    }
    return $hash;
}

package MYLWPUserAgent;

use base 'LWP::UserAgent';

sub get_basic_credentials
{
    my ($self, $realm, $uri, $proxy) = @_;
    return @{ $self->{basic_authentication}->{$uri->host_port}->{""} || [] };
}

1;
__END__

File: custisinstall/replicate.php
#!/usr/bin/php
<?php
# Автоматический скрипт репликации вики-статей
# Реплицирует как сами статьи, так и изображения, используемые в них

$PHP_SELF = array_shift($argv);
chdir(dirname($PHP_SELF));

$config = read_config(array_shift($argv));
if (!$config)
    die(<<<EOF
MediaWiki replication script by Vitaliy Filippov <vfilippov@custis.ru>

USAGE: $PHP_SELF <replication-config.ini> [targets...]
When called without target list, $PHP_SELF will attempt to replicate all targets
found in config file. There must be 2 sections in config file according to
each target and named "<Target>SourceWiki" and "<Target>DestinationWiki".

Config file fragment syntax (Replace __Test__ with desired [target] name):

[__Test__SourceWiki]
URL=<source wiki url>
Category=<source category name>
FullHistory=<'yes' or 'no' (default), 'yes' replicates all page revisions, not only the last one>
ForceImageDownload=<'yes' or 'no' (default), 'yes' means force image fetching>
BasicLogin=<HTTP basic auth username, if needed>
BasicPassword=<HTTP basic auth password, if needed>

[__Test__DestinationWiki]
URL=<destination wiki url>
Path=<destination wiki installation path>
BasicLogin=<HTTP basic auth username, if needed>
BasicPassword=<HTTP basic auth password, if needed>
User=<name of a user having import rights in destination wiki>
Password=<his password>

EOF
);

if (!function_exists('curl_init'))
    die("This script needs cURL extension to work - please enable it in your php.ini\n");

require_once "../maintenance/commandLine.inc";

$temps = array();
$targets = array_map("strtolower", $argv);
if (!count($targets))
    $targets = array_keys($config);
foreach ($targets as $target)
{
    # replicate targets
    echo "[$target] Begin replication\n";
    if ($err = replicate($config[$target][src], $config[$target][dest], $target))
        fwrite(STDERR, "[$target] Could not replicate:\n[$target] $err\n");
}
foreach ($temps as $temp)
    unlink($temp);
exit(0);

function xmlfh_repl_cb($m)
{
    global $enqa, $q;
    return $m[1].enqueue($m[2], $enqa[0], $enqa[1], $enqa[2], $q, $enqa[3], $enqa[4]).$m[3];
}

function xmlfh_trans_write($curl, $content)
{
    global $xmlfh;
    fwrite($xmlfh, preg_replace_callback('#(<src[^<>]*>)(.*?)(<\/src\s*>)#is', 'xmlfh_repl_cb', $content));
    return strlen($content);
}

function replicate($src, $dest, $targetname)
{
    global $temps, $xmlfh, $enqa, $cookiefile, $q;
    $curl = curl_init("$src[url]/index.php?title=Special:Export&action=submit");
    # Читаем список страниц категории
    curl_setopt_array($curl, array(
        CURLOPT_POST            => 1,
        CURLOPT_HEADER          => 0,
        CURLOPT_POSTFIELDS      => "addcat=Добавить&catname=$src[category]",
        CURLOPT_RETURNTRANSFER  => 1,
    ));
    if ($src[basiclogin])
        curl_setopt($curl, CURLOPT_USERPWD, $src[basiclogin].':'.$src[basicpassword]);
    $text = curl_exec($curl);
    if (!$text)
        return "Could not post '$src[url]/index.php?title=Special:Export&action=submit&catname=$src[category]': ".curl_error($curl);
    curl_close($curl);
    if (preg_match('#<textarea[^<>]*>(.*?)</textarea>#is', $text, $m))
        $text = $m[1];
    if (!$text)
        return "No pages in category $src[category]";
    $text = html_entity_decode($text);
    $ts = gettimeofday();
    $ts = $ts[sec] + $ts[usec]/1000000;
    $q = array(async => curl_multi_init());
    # Читаем экспортную XML-ку
    array_push($temps, $fn = tempnam('/tmp', 'wikiexport'));
    $fh = fopen($fn, "wb");
    if (!$fh)
        return "Could not write into temp file $fn";
    $curl = curl_init("$src[url]/index.php?title=Special:Export&action=submit");
    $xmlfh = $fh;
    $enqa = array($src[url], $dest[url], $dest[path], $src[basiclogin] ? $src[basiclogin].':'.$src[basicpassword] : '', $src[forceimagedownload]);
    curl_setopt_array($curl, array(
        CURLOPT_POST          => 1,
        CURLOPT_HEADER        => 0,
        CURLOPT_WRITEFUNCTION => 'xmlfh_trans_write',
        CURLOPT_POSTFIELDS    => "templates=1&images=1&wpDownload=1&curonly=".($src[fullhistory] ? 1 : 0)."&pages=".urlencode($text),
    ));
    if ($src[basiclogin])
        curl_setopt($curl, CURLOPT_USERPWD, $src[basiclogin].':'.$src[basicpassword]);
    $r = curl_exec($curl);
    if (!$r)
        return "Could not retrieve export XML file from '$src[url]/index.php?title=Special:Export&action=submit': ".curl_error($curl);
    curl_close($curl);
    $tx = gettimeofday();
    $tx = $tx[sec] + $tx[usec]/1000000;
    echo sprintf("[$targetname] Retrieved %d bytes in %.2f seconds\n", ftell($fh), $tx-$ts);
    fclose($fh);
    # Дожидаемся сливания картинок
    $still = 1;
    while ($still)
        curl_multi_exec($q[async], $still);
    # Публикуем картинки, вызывая код медиавики
    $total = 0;
    foreach ($q[filenames] as $base => $file)
    {
        fseek($q[fhby][$base], 0, 2);
        $tell = ftell($q[fhby][$base]);
        $total += $tell;
        fclose($q[fhby][$base]);
        curl_close($q[curl][$base]);
        if ($tell && ($src[forceimagedownload] || sha1_file($file) != $q[sha1sum][$base]))
        {
            echo "Importing $file into Image:$base...";
            # Validate a title
            $title = Title::makeTitleSafe(NS_IMAGE, $base);
            if (is_object($title))
            {
                # Check existence
                $image = wfLocalFile($title);
                if ($image->exists())
                    echo "overwriting...";
                else
                    echo "adding...";
                # Import the file
                $archive = $image->publish($file);
                if (!WikiError::isError($archive) && $archive->isGood())
                {
                    if ($image->recordUpload($archive->value, 'Imported image'))
                        echo "done.\n";
                    else
                        echo "failed: could not log upload.\n";
                }
                else
                    echo "failed: could not publish.\n";
            }
            else
                echo "failed: invalid title.\n";
        }
    }
    $ti = gettimeofday();
    $ti = $ti[sec] + $ti[usec]/1000000;
    print sprintf("[$targetname] Retrieved %d objects (total %d bytes) in %.2f seconds\n", count($q[filenames]), $total, $ti-$tx);
    curl_multi_close($q[async]);
    # Логинимся по назначению, если надо
    array_push($temps, $cookiefile = tempnam('/tmp', 'cookie'));
    if ($dest[user] && $dest[password])
    {
        $curl = curl_init("$dest[url]/index.php?title=Special:UserLogin&action=submitlogin&type=login");
        curl_setopt_array($curl, array(
            CURLOPT_POST            => 1,
            CURLOPT_HEADER          => 0,
            CURLOPT_POSTFIELDS      => "wpName=$dest[user]&wpPassword=$dest[password]&wpLoginAttempt=1",
            CURLOPT_RETURNTRANSFER  => 1,
            CURLOPT_COOKIEFILE      => $cookiefile,
            CURLOPT_COOKIEJAR       => $cookiefile,
        ));
        if ($dest[basiclogin])
            curl_setopt($curl, CURLOPT_USERPWD, $dest[basiclogin].':'.$dest[basicpassword]);
        $r = curl_exec($curl);
        if (($code = curl_getinfo($curl, CURLINFO_HTTP_CODE)) != 302)
            return "Could not login into destination wiki under user '$dest[user]': HTTP $code";
        curl_close($curl);
    }
    # Запускаем импорт исправленной XML-ки
    $curl = curl_init("$dest[url]/index.php?title=Special:Import&action=submit");
    curl_setopt_array($curl, array(
        CURLOPT_POST            => 1,
        CURLOPT_HEADER          => 0,
        CURLOPT_POSTFIELDS      => array(source => 'upload', xmlimport => '@'.$fn),
        CURLOPT_RETURNTRANSFER  => 1,
        CURLOPT_COOKIEFILE      => $cookiefile,
        CURLOPT_COOKIEJAR       => $cookiefile,
    ));
    if ($dest[basiclogin])
        curl_setopt($curl, CURLOPT_USERPWD, $dest[basiclogin].':'.$dest[basicpassword]);
    $r = curl_exec($curl);
    if (!$r)
        return "Could not import XML data into '$dest[url]/index.php?title=Special:Import&action=submit': ".curl_error($curl);
    curl_close($curl);
    if (preg_match('/<p[^<>]*class\s*=\s*["\']?error[^<>]*>\s*(.*?)\s*<\/p\s*>/is', $r, $m))
        return "Could not import XML data into $dest[url]: $m[1]";
    $tp = gettimeofday();
    $tp = $tp[sec] + $tp[usec]/1000000;
    print sprintf("[$targetname] Imported in %.2f seconds\n", $tp-$ti);
    # Всё ОК
    return '';
}

function enqueue ($url, $wikiurl, $towiki, $topath, &$q, $auth, $force)
{
    global $cookiefile, $temps;
    if (substr($url, 0, strlen($wikiurl)) == $wikiurl)
        $url = substr($url, strlen($wikiurl));
    $url = preg_replace('#^/*#s', '', $url);
    # Уже поставлено в очередь?
    if ($q[alr][$url])
        return $towiki.'/'.$url;
    $q[alr][$url] = 1;
    $fn = "$topath/$url";
    $fn = urldecode($fn);
    # архивные картинки вызывают странные баги в MediaWiki, так что пропускаем их.
    # - баги типа Fatal error: Cannot redeclare wfspecialupload() (previously
    # declared in /home/www/localhost/WWW/wiki/includes/specials/SpecialUpload.php:12)
    # in /home/www/localhost/WWW/wiki/includes/specials/SpecialUpload.php on line 15
    if (strstr($fn, '!'))
        return $towiki.'/'.$url;
    $curl = curl_init("$wikiurl/$url");
    array_push($temps, $tmpfn = tempnam('/tmp', 'img'));
    $tmpfh = fopen($tmpfn, "wb");
    curl_setopt_array($curl, array(
        CURLOPT_HEADER      => 0,
        CURLOPT_COOKIEFILE  => $cookiefile,
        CURLOPT_COOKIEJAR   => $cookiefile,
        CURLOPT_FILE        => $tmpfh,
    ));
    if ($auth)
        curl_setopt($curl, CURLOPT_USERPWD, $auth);
    # Чтобы не перезасасывать неизменённые файлы
    $stat = stat($fn);
    $sha1 = '';
    if ($stat[9])
    {
        $sha1 = sha1_file($fn);
        if (!$force)
            curl_setopt($curl, CURLOPT_HTTPHEADER, array("If-Modified-Since: ".gmdate('D, d M Y H:i:s \G\M\T', $stat[9])));
    }
    curl_multi_add_handle($q[async], $curl);
    $fn = preg_replace('#^.*/#s', '', $fn);
    $q[filenames][$fn] = $tmpfn;
    $q[sha1sum][$fn] = $sha1;
    $q[fhby][$fn] = $tmpfh;
    $q[curl][$fn] = $curl;
    return $towiki.'/'.$url;
}

function read_config ($file)
{
    if (!($fh = fopen($file, "rb")))
        return false;
    $cfg = array();
    $th = array();
    while ($s = fgets($fh))
    {
        $s = preg_replace('/(^|\s+)(;|\#).*$/s', '', preg_replace('/\s+$/s', '', $s));
        if (!$s)
            continue;
        if (preg_match('/^\s*\[([^\]]*)(Source|Destination)Wiki\]\s*$/is', $s, $m))
        {
            $target = strtolower($m[1]);
            $key = strtolower($m[2]) == 'source' ? 'src' : 'dest';
            if (is_array($cfg[$target][strtolower($m[2]) == 'source' ? 'dest' : 'src']))
                $th[$target] = 1;
        }
        else if (preg_match('/^\s*([^=]*[^\s=])\s*=\s*(.*)/is', $s, $m))
        {
            $k = strtolower($m[1]);
            $v = $m[2];
            if ($k == 'url' || $k == 'path')
                $v = preg_replace('#/+$#s', '', $v);
            else if ($k == 'fullhistory' || $k == 'forceimagedownload')
            {
                $v = strtolower($v);
                $v = $v == 'yes' || $v == 'true' || $v == 'on' || $v == '1';
            }
            if ($target && $key)
                $cfg[$target][$key][$k] = $v;
            else
                $cfg[$k] = $v;
        }
    }
    fclose($fh);
    foreach (array_keys($cfg) as $target)
        if (!$th[$target])
            unset($cfg[$target]);
    if (!count(array_keys($cfg)))
        return false;
    return $cfg;
}
